packages/timeout/src/index.ts

```diff
@@ -0,0 +1,111 @@
+
+/**
+ * Custom error class that will be used for timeout error.
+ */
+export class TimeoutError extends Error {
+}
+
+/**
+ * Custom error class to handle `tryCancel()` checks.
+ * This should not be exposed to user land, as it will be caught in.
+ */
+class InternalTimeoutError extends TimeoutError {
+}
+
+/**
+ * Checks whether we are inside timeout handler created by this package, and cancels current
+ * task execution by throwing `TimeoutError`. This error will be ignored if the promise timed
+ * out already, or explicitly skipped in `addTimeoutToPromise`.
+ *
+ * Use this function after every async call that runs inside the timeout handler:
+ *
+ * ```ts
+ * async function doSomething() {
+ *     await doSomethingTimeConsuming();
+ *     tryCancel();
+ *     await doSomethingElse();
+ *     tryCancel();
+ * }
+ * ```
+ */
+export function tryCancel(): void {
+    const { signal } = storage.getStore()?.cancelTask ?? {};
+
+    if (signal?.aborted) {
+        throw new InternalTimeoutError('Promise handler has been canceled due to a timeout');
+    }
+}
+
+/**
+ * Runs given handler and rejects with the given `errorMessage` (or `Error` instance)
+ * after given `timeoutMillis`, unless the original promise resolves or rejects earlier.
+ * Use `tryCancel()` function inside the handler after each await to finish its execution
+ * early when the timeout appears.
+ *
+ * ```ts
+ * const res = await addTimeoutToPromise(
+ *   () => handler(),
+ *   200,
+ *   'Handler timed out after 200ms!',
+ * );
+ * ```
+ */
+export async function addTimeoutToPromise<T>(handler: () => Promise<T>, timeoutMillis: number, errorMessage: string | Error): Promise<T> {
+    // respect existing context to support nesting
+    const context = storage.getStore() ?? {
+        cancelTimeout: new AbortController(),
+        cancelTask: new AbortController(),
+    };
+    let ret: T;
+
+    // calls handler, skips internal `TimeoutError`s that might have been thrown
+    // via `tryCancel()` and aborts the timeout promise after the handler finishes
+    const wrap = async () => {
+        try {
+            ret = await handler();
+        } catch (e) {
+            if (!(e instanceof InternalTimeoutError)) {
+                throw e;
+            }
+        } finally {
+            context.cancelTimeout.abort();
+        }
+    };
+
+    const timeout = async () => {
+        try {
+            await setTimeout(timeoutMillis, undefined, { signal: context.cancelTimeout.signal });
+            context.cancelTask.abort();
+        } catch (e) {
+            // ignore rejections (task finished and timeout promise has been cancelled)
+            return;
+        }
+
+        throw errorMessage instanceof Error
+            ? errorMessage
+            : new TimeoutError(errorMessage);
+    };
+
+    await new Promise((resolve, reject) => {
+        storage.run(context, () => {
+            Promise.race([timeout(), wrap()]).then(resolve, reject);
```

Comment:
We used to do timeouts with `Promise.race()` and ended up with memory leaks, because even though the timeout fired and the race resolved, the other promise kept existing and it often held a lot of data, e.g. full page HTML. Please make sure this will not happen here.

Reply:
I dont think it will happen as we are explicitly aborting them (also I can see issues with the eager promise execution which is not present here). I did test this locally directly with SDK and puppeteer crawler. Will do more extensive testing once its handled in SDK master.

Reply:
> We used to do timeouts with Promise.race() and ended up with memory leaks, because even though the timeout fired and the race resolved, the other promise kept existing and it often held a lot of data

This promise has no right to exist and should be collected by garbage collector. AFAIK, hanging promises are generally safe in JS. If it leaks then probably it's a bug in V8.

Reply:
See this issue https://github.com/apify/apify-js/issues/263 Not sure if it's still relevant, but at that time, it was an absolutely critical fix.